### PR TITLE
feat/dijkstra

### PR DESCRIPTION
**Pull request by 0xTitan**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Add a new feature in searching workspace : dijkstra

Issue Number: N/A

## What is the new behavior?

Dijkstra searching algorithm using priority queue. It finds the shortest path from a source node to all other nodes in a weighted graph.
To use it, first add many edges (nodes) to a graph using : GraphTrait::add_edge(ref graph, source, dest, weight);
Once done call GraphTrait::shortest_path(ref graph, source);
It will return a dictionnary Felt252Dict<u128> with the shortest path from source to every graph nodes.

## Does this introduce a breaking change?

- [ ] Yes
- [X ] No

## Other information

NA

### CODE REVIEW

File: src/searching/src/dijkstra.cairo [hunk 1]
```diff
@@ -0,0 +1,202 @@
+//! Dijkstra algorithm using priority queue
+use core::traits::Index;
+use core::option::OptionTrait;
+use core::array::SpanTrait;
+use core::box::BoxTrait;
+use array::{Array, ArrayTrait};
+use traits::Into;
+use dict::Felt252DictTrait;
+use nullable::{NullableTrait, nullable_from_box, match_nullable, FromNullableResult};
+use alexandria_data_structures::vec::{Felt252Vec, VecTrait};
```
Comment (LucasLvy) [inline src/searching/src/dijkstra.cairo:10]: where is that used ?

File: src/searching/src/dijkstra.cairo [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                is_null = true;
+                nodes.append(node);
+                nodes.span()
+            },
+            FromNullableResult::NotNull(adj_nodes) => {
+                adj_nodes.unbox()
+            },
+        };
+
+        //iterate over existing array to add new node
```
Comment (LucasLvy) [inline src/searching/src/dijkstra.cairo:66]: Format all the comments plz
```suggestion
        // iterate over existing array to add new node
```

File: src/searching/src/dijkstra.cairo [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            break;
+        }
+
+        let mut edge_distance: u128 = 0;
+        let mut new_distance: u128 = 0;
+        let adj_nodes = self.adj_nodes.get(visited.into());
+
+        //retrieve adj node
+        let mut adj_nodes_list = match match_nullable(adj_nodes) {
+            FromNullableResult::Null(()) => {
```
Comment (LucasLvy) [inline src/searching/src/dijkstra.cairo:130]:
```suggestion
            FromNullableResult::Null => {
```

File: src/searching/src/tests/dijkstra_test.cairo [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    GraphTrait::add_edge(ref graph, source, dest + 2, weight + 2);
+    GraphTrait::add_edge(ref graph, source, dest + 3, weight + 3);
+    GraphTrait::add_edge(ref graph, 2, 1, 2);
+    GraphTrait::add_edge(ref graph, 2, 3, 3);
+
+    assert(graph.nodes.len() == 6, 'wrong node number');
+    let val = graph.adj_nodes.get(source.into());
+
+    let span = match match_nullable(val) {
+        FromNullableResult::Null(()) => {
```
Comment (LucasLvy) [inline src/searching/src/tests/dijkstra_test.cairo:31]:
```suggestion
        FromNullableResult::Null => {
```

File: src/searching/src/tests/dijkstra_test.cairo [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    assert(new_node.weight == weight + 1, 'Wrong weight in adj edge');
+
+    let new_node = *span.get(3).unwrap().unbox();
+    assert(new_node.dest == dest + 3, 'Wrong dest in adj edge');
+    assert(new_node.weight == weight + 3, 'Wrong weight in adj edge');
+
+    let val = graph.adj_nodes.get(2.into());
+
+    let span = match match_nullable(val) {
+        FromNullableResult::Null(()) => {
```
Comment (LucasLvy) [inline src/searching/src/tests/dijkstra_test.cairo:52]:
```suggestion
        FromNullableResult::Null => {
```

File: src/searching/src/dijkstra.cairo [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        Graph { nodes: ArrayTrait::new(), adj_nodes: Default::default() }
+    }
+
+    fn add_edge(ref self: Graph<Nullable<Span<Node>>>, source: u32, dest: u32, weight: u128) {
+        let adj_nodes = self.adj_nodes.get(source.into());
+        let mut nodes: Array<Node> = ArrayTrait::new();
+        let mut is_null: bool = false;
+        let node = Node { source, dest, weight };
+        let mut span = match match_nullable(adj_nodes) {
+            FromNullableResult::Null(()) => {
```
Comment (LucasLvy) [inline src/searching/src/dijkstra.cairo:56]:
```suggestion
            FromNullableResult::Null => {
```

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: src/searching/src/dijkstra.cairo (hunk 1 above)
Comment (0xTitan) [reply src/searching/src/dijkstra.cairo]: Hello, thx a lot for the review.

- use traits::Into; is needed like in line 50.
- use dict::Felt252DictTrait; is used to call insert, squash, get on dictionnary. See Graph struct.
- I rework use nullable::{NullableTrait, nullable_from_box, match_nullable, FromNullableResult}; to only import use nullable::{FromNullableResult};
- use alexandria_data_structures::vec::{Felt252Vec, VecTrait}; is removed

Let me know if you're think is not ok.

Comment (LucasLvy) [reply src/searching/src/dijkstra.cairo]: perfect

File: src/searching/src/tests/dijkstra_test.cairo [hunk 7]
```diff
@@ -0,0 +1,129 @@
+use core::debug::PrintTrait;
+use core::array::SpanTrait;
+use core::box::BoxTrait;
+use traits::Into;
+use option::OptionTrait;
+use array::{Array, ArrayTrait};
+use dict::Felt252DictTrait;
+use nullable::{FromNullableResult};
```
Comment (LucasLvy) [inline src/searching/src/tests/dijkstra_test.cairo:8]:
```suggestion
use nullable::FromNullableResult;
```

File: src/searching/src/dijkstra.cairo [hunk 8]
```diff
@@ -0,0 +1,200 @@
+//! Dijkstra algorithm using priority queue
+use core::option::OptionTrait;
+use core::array::SpanTrait;
+use core::box::BoxTrait;
+use array::{Array, ArrayTrait};
+use traits::{Into, Index};
+use dict::Felt252DictTrait;
+use nullable::{FromNullableResult};
```
Comment (LucasLvy) [inline src/searching/src/dijkstra.cairo:8]:
```suggestion
use nullable::FromNullableResult;
```

File: src/searching/src/tests/dijkstra_test.cairo [hunk 9]
```diff
@@ -0,0 +1,129 @@
+use core::debug::PrintTrait;
+use core::array::SpanTrait;
+use core::box::BoxTrait;
```
Comment (LucasLvy) [inline src/searching/src/tests/dijkstra_test.cairo:3]:
```suggestion
use debug::PrintTrait;
use box::BoxTrait;
```

File: src/searching/src/dijkstra.cairo [hunk 10]
```diff
@@ -0,0 +1,200 @@
+//! Dijkstra algorithm using priority queue
+use core::option::OptionTrait;
+use core::array::SpanTrait;
+use core::box::BoxTrait;
```
Comment (LucasLvy) [inline src/searching/src/dijkstra.cairo:4]:
```suggestion
use option::OptionTrait;
use box::BoxTrait;
```

File: src/searching/src/tests/dijkstra_test.cairo [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+use traits::Into;
+use option::OptionTrait;
+use array::{Array, ArrayTrait};
```
Comment (LucasLvy) [inline src/searching/src/tests/dijkstra_test.cairo:6]:
```suggestion
use array::{Array, ArrayTrait, SpanTrait};
```

File: src/searching/src/dijkstra.cairo [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+use array::{Array, ArrayTrait};
```
Comment (LucasLvy) [inline src/searching/src/dijkstra.cairo:5]:
```suggestion
use array::{Array, ArrayTrait, SpanTrait};
```

**Review by LucasLvy — CHANGES_REQUESTED** [review]

**Review by LucasLvy — APPROVED** [review]

**Pull request merged**